packages/firestore/src/remote/bloom_filter.ts

```diff
@@ -107,4 +118,40 @@ export class BloomFilter {
     }
     return true;
   }
+
+  /** Create bloom filter input for testing purposes only. */
+  static create(
+    numOfBits: number,
+    hashCount: number,
+    contains: string[]
+  ): BloomFilter {
+    const padding = numOfBits % 8 === 0 ? 0 : 8 - (numOfBits % 8);
+    const bitmap = new Uint8Array(Math.ceil(numOfBits / 8));
+    const bloomFilter = new BloomFilter(bitmap, padding, hashCount);
+    contains.forEach(item => bloomFilter.insert(item));
+    return bloomFilter;
+  }
+
+  private insert(value: string): void {
+    if (this.size === 0 || value === '') {
```

Comment:
But mightContain can handle '', and it always return false. So we insert an empty string '', which does nothing, and wouldn't stop us from inserting other keys. And call mightContain on the bitmap created, empty string returns false.

hmm, I am seeing the flaw here, insert empty string and return false for mightContain(''). I will change it to throw error.

Reply:
> insert empty string and return false for mightContain('')

Exactly. But this actually makes me think of something else: why do we prevent inserting the empty string at all? I don't see any reason why it would be problematic because the empty string can be hashed just like any other string. I'm 99% sure that it was me that said we should handle the empty string as a special case, but now I think I was wrong.

For now, let's leave this special treatment in place as you propose. But can you create a taskflow task to re-visit this decision? I'm now thinking that we should _not_ treat the empty string specially but rather treat it the same as any other string: hash it and test for its membership.

Reply:
By returning false early on empty bloom filter and empty string, we saved the time we would have used on doing all the hashing and calculating for nothing.

Reply:
I think we should _still_ unconditionally return `false` on an empty bloom filter; however, we should consider that the bloom filter _can_ contain the empty string, and test for its membership like any other string. Basically, we should remove the artificial restriction that the bloom filter cannot contain the empty string.

Reply:
With current implementation, we cannot do that cause the md5 hash on empty string returns normal 16 byte array.  
<img width="258" alt="Screenshot 2023-01-17 at 5 50 27 PM" src="https://user-images.githubusercontent.com/107142260/213061530-b61153ee-4d00-4c4b-b309-2760442ac759.png">

So as long as the bitmap is not empty, there is a high chance that empty string membership returns true. 

Unless, we specifically require md5 hash don't hash empty strings, and still end up ruling out empty string.

Reply:
That is okay. What I am suggesting is that we _allow_ the empty string to be contained in the bloom filter, just like we allow any other string. And just like any other string, there is a non-zero probability that it will indeed be contained in the bloom filter by coincidence, a false positive. That is okay, and expected. I'm just saying there is no need for us to special case the empty string. It's true that with this change calling `mightContain('')` could erroneously return true, and that's okay because false positives are expected.